File: app/src/main/java/com/simon/harmonichackernews/utils/CommentSorter.java
package com.simon.harmonichackernews.utils;

import android.content.Context;

import com.simon.harmonichackernews.data.Comment;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

public class CommentSorter {

    public static void sort(Context ctx, List<Comment> comments) {
        sort(comments, SettingsUtils.getPreferredCommentSorting(ctx));
    }

    public static void sort(List<Comment> comments, String sortType) {
        switch (sortType) {
            case "Default":
                sortComments(comments, new Comparator<Comment>() {
                    @Override
                    public int compare(Comment c1, Comment c2) {
                        return Integer.compare(c1.sortOrder, c2.sortOrder);
                    }
                }, false);
                break;
            case "Reply count":
                sortComments(comments, new Comparator<Comment>() {
                    @Override
                    public int compare(Comment c1, Comment c2) {
                        return Integer.compare(c2.totalReplies, c1.totalReplies);
                    }
                }, true);
                break;
            case "Newest first":
                sortComments(comments, new Comparator<Comment>() {
                    @Override
                    public int compare(Comment c1, Comment c2) {
                        return Integer.compare(c2.time, c1.time);
                    }
                }, false);
                break;
            case "Oldest first":
                sortComments(comments, new Comparator<Comment>() {
                    @Override
                    public int compare(Comment c1, Comment c2) {
                        return Integer.compare(c1.time, c2.time);
                    }
                }, false);
                break;
        }
    }

    private static void sortComments(List<Comment> comments, Comparator<Comment> comparator, boolean updateReplyCounts) {
        if (comments.size() <= 1) {
            return;
        }

        Comment header = comments.get(0);
        List<Comment> commentsWithChildren = buildCommentTree(comments);

        if (updateReplyCounts) {
            updateTotalReplies(commentsWithChildren);
        }

        sortCommentsRecursive(commentsWithChildren, comparator);

        comments.clear();
        comments.add(header);
        flattenComments(commentsWithChildren, comments);
    }

    private static void sortCommentsRecursive(List<Comment> commentsWithChildren, Comparator<Comment> comparator) {
        Collections.sort(commentsWithChildren, comparator);

        for (Comment c : commentsWithChildren) {
            sortCommentsRecursive(c.childComments, comparator);
        }
    }


    private static List<Comment> buildCommentTree(List<Comment> comments) {
        List<Comment> commentsWithChildren = new ArrayList<>();
        List<Comment> parentsByDepth = new ArrayList<>();

        for (int i = 1; i < comments.size(); i++) {
            Comment comment = comments.get(i);
            comment.childComments = new ArrayList<>();
            int depth = Math.max(0, comment.depth);

            while (parentsByDepth.size() > depth) {
                parentsByDepth.remove(parentsByDepth.size() - 1);
            }

            if (depth == 0 || parentsByDepth.isEmpty()) {
                commentsWithChildren.add(comment);
            } else {
                parentsByDepth.get(parentsByDepth.size() - 1).childComments.add(comment);
            }

            parentsByDepth.add(comment);
        }

        return commentsWithChildren;
    }

    private static void flattenComments(List<Comment> comments, List<Comment> flatComments) {
        for (Comment comment : comments) {
            flatComments.add(comment);
            if (!comment.childComments.isEmpty()) {
                flattenComments(comment.childComments, flatComments);
            }
        }
    }

    private static void updateTotalReplies(List<Comment> comments) {
        for (Comment comment : comments) {
            updateTotalReplies(comment);
        }
    }

    private static int updateTotalReplies(Comment comment) {
        int count = 0;
        for (Comment child : comment.childComments) {
            count += 1 + updateTotalReplies(child);
        }
        comment.totalReplies = count;
        return count;
    }

}

File: app/src/main/java/com/simon/harmonichackernews/utils/CommentSorter.kt
package com.simon.harmonichackernews.utils

import android.content.Context
import com.simon.harmonichackernews.data.Comment
import kotlin.math.max

object CommentSorter {
    fun sort(ctx: Context, comments: MutableList<Comment>) {
        sort(comments, SettingsUtils.getPreferredCommentSorting(ctx))
    }

    fun sort(comments: MutableList<Comment>, sortType: String) {
        when (sortType) {
            "Default" -> {
                if (isInDefaultOrder(comments)) return
                sortComments(comments, compareBy(Comment::sortOrder), false)
            }

            "Reply count" -> sortComments(
                comments,
                compareByDescending(Comment::totalReplies),
                true,
            )

            "Newest first" -> sortComments(
                comments,
                compareByDescending(Comment::time),
                false,
            )

            "Oldest first" -> sortComments(comments, compareBy(Comment::time), false)
        }
    }

    private fun isInDefaultOrder(comments: List<Comment>): Boolean =
        comments.size < 3 || comments.subList(1, comments.size).isSortedBy(Comment::sortOrder)

    private fun sortComments(
        comments: MutableList<Comment>,
        comparator: Comparator<Comment>,
        updateReplyCounts: Boolean
    ) {
        if (comments.size <= 1) return

        val header = comments.first()
        val commentsWithChildren = buildCommentTree(comments)

        if (updateReplyCounts) {
            updateTotalReplies(commentsWithChildren)
        }

        sortCommentsRecursive(commentsWithChildren, comparator)

        comments.clear()
        comments.add(header)
        flattenComments(commentsWithChildren, comments)
    }

    private fun sortCommentsRecursive(
        commentsWithChildren: MutableList<Comment>,
        comparator: Comparator<Comment>
    ) {
        commentsWithChildren.sortWith(comparator)

        for (comment in commentsWithChildren) {
            sortCommentsRecursive(comment.childComments, comparator)
        }
    }

    private fun buildCommentTree(comments: List<Comment>): MutableList<Comment> {
        val commentsWithChildren = mutableListOf<Comment>()
        val parentsByDepth = mutableListOf<Comment>()

        for (i in 1..<comments.size) {
            val comment = comments[i]
            comment.childComments = mutableListOf()
            val depth = max(0, comment.depth)

            while (parentsByDepth.size > depth) {
                parentsByDepth.removeAt(parentsByDepth.lastIndex)
            }

            if (depth == 0 || parentsByDepth.isEmpty()) {
                commentsWithChildren.add(comment)
            } else {
                parentsByDepth.last().childComments.add(comment)
            }

            parentsByDepth.add(comment)
        }

        return commentsWithChildren
    }

    private fun flattenComments(
        comments: MutableList<Comment>,
        flatComments: MutableList<Comment>
    ) {
        for (comment in comments) {
            flatComments.add(comment)
            if (comment.childComments.isNotEmpty()) {
                flattenComments(comment.childComments, flatComments)
            }
        }
    }

    private fun updateTotalReplies(comments: MutableList<Comment>) {
        for (comment in comments) {
            updateTotalReplies(comment)
        }
    }

    private fun updateTotalReplies(comment: Comment): Int {
        val count = comment.childComments.sumOf { child -> 1 + updateTotalReplies(child) }
        comment.totalReplies = count
        return count
    }
}
